app/models/course/material/folder.rb

```diff
@@ -1,7 +1,17 @@
 class Course::Material::Folder < ActiveRecord::Base
-  belongs_to :parent_folder, inverse_of: :subfolders, class_name: Course::Material::Folder.name
-  has_many :subfolders, inverse_of: :parent_folder, dependent: :destroy,
-                        foreign_key: :parent_folder_id, class_name: Course::Material::Folder.name
+  acts_as_forest order: 'name'
+
+  after_initialize :set_defaults, if: :new_record?
+
   has_many :materials, inverse_of: :folder, dependent: :destroy, foreign_key: :folder_id,
                        class_name: Course::Material.name
+  belongs_to :course, inverse_of: :material_folders
+
+  validates :name, uniqueness: { case_sensitive: false, scope: :parent_id }, if: :parent
```

Comment:
This should be done at the database, schema_plus_validations should be able to infer this.